test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -60,6 +85,27 @@
         ;; make sure we got here
         (is (true? true))))))
 
+(deftest pool-lock-is-blocking-test-until-borrows-unregistered-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 2)
+        future-started? (promise)]
+    (is (= 2 (count instances)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool))]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are unregistered"
+        (is (not (realized? lock-thread)))
+
+        (testing (str "other threads may successfully unregister instances "
+                      "while pool.lock() is being executed")
+          (.unregister pool (first instances))
+          (is (not (realized? lock-thread)))
+          (.unregister pool (second instances)))
+
+        @lock-thread
+        ;; make sure we got here
+        (is (true? true))))))
```

Comment:
The way the code works in the PR currently, the pending lock request would be granted after the last instance is unregistered.  But, then, the locker would have a lock on a pool that is now empty - at least temporarily, because the poison pill would end up being dumped into the pool right after that.  I don't have any logic in the implementation that would prevent a pill from being inserted while the pool lock is held.  Not sure if we need that.

I'm not sure I agree with the semantics of having a lock request be aborted if instances are unregistered from the pool while the lock request is still pending.  Seems like it still makes sense for the locker to get access to the pool in that case.

This may be bleeding over into the question I'd asked you today about what could happen if a file sync write lock were to occur right around the time that a full JRuby pool flush is being performed.  I think it might be better to deal with that question separately from this PR.  I'll file a separate ticket for that and link that back to this PR.  Sound okay?